tests/logging/test_tracing.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2022 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from twisted.internet import defer
+from twisted.test.proto_helpers import MemoryReactorClock
+
+from synapse.logging.context import make_deferred_yieldable, run_in_background
+from synapse.logging.tracing import start_active_span
+from synapse.util import Clock
+
+from tests.unittest import TestCase
+
+try:
+    import opentelemetry
+    import opentelemetry.sdk.trace
+    import opentelemetry.sdk.trace.export
+    import opentelemetry.sdk.trace.export.in_memory_span_exporter
+    import opentelemetry.trace
+    import opentelemetry.trace.propagation
+except ImportError:
+    opentelemetry = None  # type: ignore[assignment]
+
+
+class TracingTestCase(TestCase):
+    """
+    Test logging contexts and active opentelemetry spans.
+    """
+
+    if opentelemetry is None:
+        skip = "Requires opentelemetry"  # type: ignore[unreachable]
+
+    def setUp(self) -> None:
+        # since this is a unit test, we don't really want to mess around with the
+        # global variables that power opentracing. We create our own tracer instance
+        # and test with it.
+
+        self._tracer_provider = opentelemetry.sdk.trace.TracerProvider()
+
+        self._exporter = (
+            opentelemetry.sdk.trace.export.in_memory_span_exporter.InMemorySpanExporter()
+        )
+        processor = opentelemetry.sdk.trace.export.SimpleSpanProcessor(self._exporter)
+        self._tracer_provider.add_span_processor(processor)
+
+        self._tracer = self._tracer_provider.get_tracer(__name__)
+
+    def test_start_active_span(self) -> None:
+        # This means no current span
+        self.assertEqual(
+            opentelemetry.trace.get_current_span(), opentelemetry.trace.INVALID_SPAN
+        )
+
+        # start_active_span should start and activate a span.
+        with start_active_span("new-span", tracer=self._tracer) as span:
+            self.assertEqual(opentelemetry.trace.get_current_span(), span)
+            self.assertIsNotNone(span.start_time)  # type: ignore[attr-defined]
```

Comment:
Is there a better way than these type ignores? Added for these `mypy` errors:

```
+ mypy
tests/logging/test_tracing.py:67: error: "Span" has no attribute "start_time"  [attr-defined]
tests/logging/test_tracing.py:73: error: "Span" has no attribute "end_time"  [attr-defined]
tests/logging/test_tracing.py:95: error: "Span" has no attribute "parent"  [attr-defined]
tests/logging/test_tracing.py:105: error: "Span" has no attribute "parent"  [attr-defined]
tests/logging/test_tracing.py:111: error: "Span" has no attribute "end_time"  [attr-defined]
tests/logging/test_tracing.py:112: error: "Span" has no attribute "end_time"  [attr-defined]
Found 6 errors in 1 file (checked 722 source files)
```

The tests pass and the attributes exist but `opentelemetry` doesn't use the types we would need and this seems discouraged in general, https://github.com/open-telemetry/opentelemetry-python/issues/1001. Is there a different way to assert things?

 - API `opentelemetry.trace.Tracer.start_span()`, https://opentelemetry-python.readthedocs.io/en/latest/api/trace.html#opentelemetry.trace.Tracer.start_span
    - Returns `opentelemetry.trace.span.Span`
 - SDK `opentelemetry.sdk.trace.Tracer.start_span()`, https://opentelemetry-python.readthedocs.io/en/latest/sdk/trace.html#opentelemetry.sdk.trace.Tracer.start_span
    - Returns `opentelemetry.trace.span.Span`

`opentelemetry.trace.Span` is also referenced some places but this just looks like it's the import of [`opentelemetry.trace.span.Span`](https://github.com/open-telemetry/opentelemetry-python/blob/c6d7b9f3dd4864445050f86e1e5a88100f1eb84e/opentelemetry-api/src/opentelemetry/trace/span.py#L57) which doesn't have the properties

[`opentelemetry.sdk.trace.Span`](https://github.com/open-telemetry/opentelemetry-python/blob/c6d7b9f3dd4864445050f86e1e5a88100f1eb84e/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L701) has these properties because it inherits from `ReadableSpan`

Reply:
We don't really need to check `span.end_time` because any span that is ended will be in the `exporter.get_finished_spans()` we already look at. Same with `start_time`, if it's ended, it also has a start time.

For the `parent` stuff, just going to opt to remove it. We could probably look at the `ReadableSpan` from the `exporter.get_finished_spans()` but probably fine not worrying about it.

🤷 These tests in general are kinda just testing the OpenTelemetry internals but it's a bit good to make sure our `start_active_span` is doing something.